File: CoinToss.cs
# CoinToss.cs
/**
 * 🪙 Coin Toss – Advanced Flip Series Analyzer (C# Edition)
 * Features: biased probability, streaks, ASCII chart, export, colorful CLI
 * Requires: .NET 6.0+
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

// ─── Core Logic ──────────────────────────────────────────────────────────────

class CoinToss
{
    private readonly double prob;
    private readonly Random random = new();

    public CoinToss(double prob = 0.5) => this.prob = prob;

    public string Flip() => random.NextDouble() < prob ? "H" : "T";

    public List<string> FlipSeries(int n)
    {
        var res = new List<string>(n);
        for (int i = 0; i < n; i++) res.Add(Flip());
        return res;
    }
}

// ─── Statistics ─────────────────────────────────────────────────────────────

class Stats
{
    public int Heads { get; set; }
    public int Tails { get; set; }
    public int Total { get; set; }
    public double HeadsPercent { get; set; }
    public double TailsPercent { get; set; }
    public int LongestRunHeads { get; set; }
    public int LongestRunTails { get; set; }
    public int LongestStreakHeads { get; set; }
    public int LongestStreakTails { get; set; }
    public int Runs { get; set; }
}

static class Analyzer
{
    public static Stats Analyze(List<string> series)
    {
        if (series.Count == 0) return new Stats();
        int heads = series.Count(s => s == "H");
        int total = series.Count;
        int tails = total - heads;
        int longestRunHeads = 0, longestRunTails = 0;
        int currentRun = 1;
        for (int i = 1; i < series.Count; i++)
        {
            if (series[i] == series[i-1]) currentRun++;
            else
            {
                if (series[i-1] == "H") longestRunHeads = Math.Max(longestRunHeads, currentRun);
                else longestRunTails = Math.Max(longestRunTails, currentRun);
                currentRun = 1;
            }
        }
        if (series.Count > 0)
        {
            if (series[^1] == "H") longestRunHeads = Math.Max(longestRunHeads, currentRun);
            else longestRunTails = Math.Max(longestRunTails, currentRun);
        }
        int runs = 1;
        for (int i = 1; i < series.Count; i++)
            if (series[i] != series[i-1]) runs++;
        return new Stats
        {
            Heads = heads,
            Tails = tails,
            Total = total,
            HeadsPercent = total > 0 ? (double)heads / total * 100 : 0,
            TailsPercent = total > 0 ? (double)tails / total * 100 : 0,
            LongestRunHeads = longestRunHeads,
            LongestRunTails = longestRunTails,
            LongestStreakHeads = longestRunHeads,
            LongestStreakTails = longestRunTails,
            Runs = runs
        };
    }

    public static string DrawAsciiChart(Stats stats, int width = 40)
    {
        int hBar = (int)(stats.HeadsPercent / 100 * width);
        int tBar = (int)(stats.TailsPercent / 100 * width);
        string line = $"H {ColorString(new string('█', hBar), "green")}{new string(' ', width - hBar)} {stats.HeadsPercent:F1}%\n";
        line += $"T {ColorString(new string('█', tBar), "red")}{new string(' ', width - tBar)} {stats.TailsPercent:F1}%";
        return line;
    }

    private static string ColorString(string s, string color)
    {
        string code = color switch
        {
            "green" => "\u001B[32m",
            "red" => "\u001B[31m",
            _ => "\u001B[0m"
        };
        return code + s + "\u001B[0m";
    }
}

// ─── Export ─────────────────────────────────────────────────────────────────

class Exporter
{
    public static void ExportResults(List<string> series, Stats stats, string filename = null)
    {
        if (string.IsNullOrEmpty(filename))
        {
            string ts = DateTime.Now.ToString("yyyyMMdd_HHmmss");
            filename = $"coin_toss_{ts}.json";
        }
        var data = new
        {
            flips = series,
            stats = new
            {
                stats.Heads,
                stats.Tails,
                stats.Total,
                stats.HeadsPercent,
                stats.TailsPercent,
                stats.LongestRunHeads,
                stats.LongestRunTails,
                stats.LongestStreakHeads,
                stats.LongestStreakTails,
                stats.Runs
            },
            timestamp = DateTime.Now.ToString("o")
        };
        string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(filename, json);
        Console.WriteLine(ColorString($"✅ Results exported to {filename}", "green"));
    }

    private static string ColorString(string s, string color)
    {
        string code = color switch
        {
            "green" => "\u001B[32m",
            _ => "\u001B[0m"
        };
        return code + s + "\u001B[0m";
    }
}

// ─── Main App ──────────────────────────────────────────────────────────────

class CoinTossApp
{
    private static readonly string Reset = "\u001B[0m";
    private static readonly string Bright = "\u001B[1m";
    private static readonly string Dim = "\u001B[2m";
    private static readonly string Red = "\u001B[31m";
    private static readonly string Green = "\u001B[32m";
    private static readonly string Yellow = "\u001B[33m";
    private static readonly string Cyan = "\u001B[36m";

    private static string C(string text, string color) => color + text + Reset;

    private double prob = 0.5;
    private CoinToss toss;
    private List<string> lastSeries = new();
    private Stats lastStats = new();

    public CoinTossApp() => toss = new CoinToss(prob);

    private string Ask(string prompt)
    {
        Console.Write(prompt);
        return Console.ReadLine()?.Trim() ?? "";
    }

    private int AskInt(string prompt, int def)
    {
        while (true)
        {
            string ans = Ask(prompt);
            if (string.IsNullOrEmpty(ans)) return def;
            if (int.TryParse(ans, out int val)) return val;
            Console.WriteLine(C("Please enter a number.", Yellow));
        }
    }

    private double AskDouble(string prompt, double def)
    {
        while (true)
        {
            string ans = Ask(prompt);
            if (string.IsNullOrEmpty(ans)) return def;
            if (double.TryParse(ans, out double val)) return val;
            Console.WriteLine(C("Please enter a number.", Yellow));
        }
    }

    private void ShowMenu()
    {
        Console.WriteLine("\n" + C(new string('═', 50), Cyan));
        Console.WriteLine(C("🪙 COIN TOSS ANALYZER", Bright + Cyan));
        Console.WriteLine(C(new string('═', 50), Cyan));
        Console.WriteLine($"  Probability of Heads: {prob:F2}");
        Console.WriteLine(C(new string('═', 50), Cyan));
        Console.WriteLine("  1. 🪙 Single Flip");
        Console.WriteLine("  2. 📊 Series of Flips");
        Console.WriteLine("  3. 📈 Show Statistics");
        Console.WriteLine("  4. 🎯 Set Probability");
        Console.WriteLine("  5. 💾 Export Results");
        Console.WriteLine("  6. 🔁 Continuous Flip Mode");
        Console.WriteLine("  0. 🚪 Exit");
        Console.WriteLine(C(new string('═', 50), Cyan));
    }

    private void SingleFlip()
    {
        string res = toss.Flip();
        string color = res == "H" ? Green : Red;
        Console.WriteLine($"\n🪙 Result: {C(res, color)} (Heads={prob:F2})");
        lastSeries.Add(res);
        lastStats = Analyzer.Analyze(lastSeries);
    }

    private void SeriesFlips()
    {
        int n = AskInt("Number of flips (default 10): ", 10);
        if (n <= 0) { Console.WriteLine(C("Number must be positive.", Red)); return; }
        Console.WriteLine(C($"Flipping {n} times...", Dim));
        lastSeries = toss.FlipSeries(n);
        lastStats = Analyzer.Analyze(lastSeries);
        ShowStats();
    }

    private void ShowStats()
    {
        if (lastStats.Total == 0)
        {
            Console.WriteLine(C("No data yet. Do some flips!", Yellow));
            return;
        }
        Console.WriteLine("\n📊 STATISTICS");
        Console.WriteLine(C(new string('─', 30), Dim));
        Console.WriteLine($"  Total Flips: {lastStats.Total}");
        Console.WriteLine($"  Heads: {lastStats.Heads} ({lastStats.HeadsPercent:F1}%)");
        Console.WriteLine($"  Tails: {lastStats.Tails} ({lastStats.TailsPercent:F1}%)");
        Console.WriteLine($"  Longest Run (H): {lastStats.LongestRunHeads}");
        Console.WriteLine($"  Longest Run (T): {lastStats.LongestRunTails}");
        Console.WriteLine($"  Number of Runs: {lastStats.Runs}");
        Console.WriteLine("\n📊 Distribution:");
        Console.WriteLine(Analyzer.DrawAsciiChart(lastStats));
    }

    private void SetProbability()
    {
        double p = AskDouble("Probability of Heads (0.0–1.0): ", 0.5);
        if (p >= 0 && p <= 1)
        {
            prob = p;
            toss = new CoinToss(p);
            Console.WriteLine(C($"✅ Probability set to {p:F2}", Green));
        }
        else Console.WriteLine(C("Invalid probability.", Red));
    }

    private void ExportResults()
    {
        if (lastStats.Total == 0)
        {
            Console.WriteLine(C("No data to export.", Yellow));
            return;
        }
        string fname = Ask("Filename (leave empty for auto): ");
        if (string.IsNullOrEmpty(fname)) fname = null;
        Exporter.ExportResults(lastSeries, lastStats, fname);
    }

    private void ContinuousFlip()
    {
        Console.WriteLine(C("\n🔁 Continuous Flip Mode (press Ctrl+C to stop)", Cyan));
        int count = 0, heads = 0, tails = 0;
        var series = new List<string>();
        while (true)
        {
            string res = toss.Flip();
            series.Add(res);
            if (res == "H") heads++; else tails++;
            count++;
            Console.Write($"\r {count} flips: H={heads} ({(double)heads/count*100:F1}%) T={tails} ({(double)tails/count*100:F1}%)");
        }
    }

    public void Run()
    {
        Console.Clear();
        Console.WriteLine(C("\n🪙 Coin Toss – Advanced Flip Analyzer", Bright + Cyan));
        Console.WriteLine(C("Flip, analyze, and master probability!", Dim));

        while (true)
        {
            ShowMenu();
            string choice = Ask("Your choice: ");
            switch (choice)
            {
                case "1": SingleFlip(); break;
                case "2": SeriesFlips(); break;
                case "3": ShowStats(); break;
                case "4": SetProbability(); break;
                case "5": ExportResults(); break;
                case "6": ContinuousFlip(); break;
                case "0":
                    Console.WriteLine(C("👋 Goodbye!", Cyan));
                    return;
                default:
                    Console.WriteLine(C("❌ Invalid choice.", Red));
                    break;
            }
            if (choice != "0" && choice != "6")
            {
                Console.Write("\nPress Enter to continue...");
                Console.ReadLine();
            }
        }
    }

    public static void Main()
    {
        try
        {
            new CoinTossApp().Run();
        }
        catch (Exception ex)
        {
            Console.WriteLine(C($"❌ Unexpected error: {ex.Message}", Red));
            Environment.Exit(1);
        }
    }
}

File: CoinToss.java
# CoinToss.java
/**
 * 🪙 Coin Toss – Advanced Flip Series Analyzer (Java Edition)
 * Features: biased probability, streaks, ASCII chart, export, colorful CLI
 * Requires: Java 17+
 */

import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.util.*;
import java.util.stream.Collectors;

// ─── Core Logic ──────────────────────────────────────────────────────────────

class CoinToss {
    private final double prob;
    private final Random random;

    public CoinToss(double prob) {
        this.prob = prob;
        this.random = new Random();
    }

    public String flip() {
        return random.nextDouble() < prob ? "H" : "T";
    }

    public List<String> flipSeries(int n) {
        List<String> res = new ArrayList<>(n);
        for (int i = 0; i < n; i++) res.add(flip());
        return res;
    }
}

// ─── Statistics ─────────────────────────────────────────────────────────────

class Stats {
    int heads, tails, total;
    double headsPercent, tailsPercent;
    int longestRunHeads, longestRunTails;
    int longestStreakHeads, longestStreakTails;
    int runs;

    Stats() {}

    Stats(int heads, int tails, int total, double headsPercent, double tailsPercent,
          int longestRunHeads, int longestRunTails, int longestStreakHeads, int longestStreakTails, int runs) {
        this.heads = heads;
        this.tails = tails;
        this.total = total;
        this.headsPercent = headsPercent;
        this.tailsPercent = tailsPercent;
        this.longestRunHeads = longestRunHeads;
        this.longestRunTails = longestRunTails;
        this.longestStreakHeads = longestStreakHeads;
        this.longestStreakTails = longestStreakTails;
        this.runs = runs;
    }
}

class Analyzer {
    static Stats analyze(List<String> series) {
        if (series.isEmpty()) return new Stats();
        int heads = (int) series.stream().filter(s -> s.equals("H")).count();
        int total = series.size();
        int tails = total - heads;
        int longestRunHeads = 0, longestRunTails = 0;
        int currentRun = 1;
        for (int i = 1; i < series.size(); i++) {
            if (series.get(i).equals(series.get(i-1))) {
                currentRun++;
            } else {
                if (series.get(i-1).equals("H")) longestRunHeads = Math.max(longestRunHeads, currentRun);
                else longestRunTails = Math.max(longestRunTails, currentRun);
                currentRun = 1;
            }
        }
        if (!series.isEmpty()) {
            if (series.get(series.size()-1).equals("H")) longestRunHeads = Math.max(longestRunHeads, currentRun);
            else longestRunTails = Math.max(longestRunTails, currentRun);
        }
        int runs = 1;
        for (int i = 1; i < series.size(); i++) {
            if (!series.get(i).equals(series.get(i-1))) runs++;
        }
        double hP = total > 0 ? (double)heads / total * 100 : 0;
        double tP = total > 0 ? (double)tails / total * 100 : 0;
        return new Stats(heads, tails, total, hP, tP,
                         longestRunHeads, longestRunTails,
                         longestRunHeads, longestRunTails, runs);
    }

    static String drawAsciiChart(Stats stats, int width) {
        int hBar = (int)(stats.headsPercent / 100 * width);
        int tBar = (int)(stats.tailsPercent / 100 * width);
        String line = "H " + colorString("█".repeat(hBar), "green") + " ".repeat(width - hBar) + " " + String.format("%.1f%%\n", stats.headsPercent);
        line += "T " + colorString("█".repeat(tBar), "red") + " ".repeat(width - tBar) + " " + String.format("%.1f%%", stats.tailsPercent);
        return line;
    }

    private static String colorString(String s, String color) {
        // Simple color mapping for cross-platform; we'll use ANSI codes
        String code = switch (color) {
            case "green" -> "\u001B[32m";
            case "red" -> "\u001B[31m";
            default -> "\u001B[0m";
        };
        return code + s + "\u001B[0m";
    }
}

// ─── Export ─────────────────────────────────────────────────────────────────

class Exporter {
    static void exportResults(List<String> series, Stats stats, String filename) throws IOException {
        if (filename == null || filename.isEmpty()) {
            String ts = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss"));
            filename = "coin_toss_" + ts + ".json";
        }
        Map<String, Object> data = new LinkedHashMap<>();
        data.put("flips", series);
        data.put("stats", Map.of(
            "heads", stats.heads,
            "tails", stats.tails,
            "total", stats.total,
            "headsPercent", stats.headsPercent,
            "tailsPercent", stats.tailsPercent,
            "longestRunHeads", stats.longestRunHeads,
            "longestRunTails", stats.longestRunTails,
            "longestStreakHeads", stats.longestStreakHeads,
            "longestStreakTails", stats.longestStreakTails,
            "runs", stats.runs
        ));
        data.put("timestamp", Instant.now().toString());
        String json = new com.google.gson.GsonBuilder().setPrettyPrinting().create().toJson(data);
        Files.writeString(Paths.get(filename), json);
        System.out.println(colorString("✅ Results exported to " + filename, "green"));
    }

    private static String colorString(String s, String color) {
        String code = switch (color) {
            case "green" -> "\u001B[32m";
            default -> "\u001B[0m";
        };
        return code + s + "\u001B[0m";
    }
}

// ─── Main App ──────────────────────────────────────────────────────────────

public class CoinTossApp {
    private static final String RESET = "\u001B[0m";
    private static final String BRIGHT = "\u001B[1m";
    private static final String DIM = "\u001B[2m";
    private static final String RED = "\u001B[31m";
    private static final String GREEN = "\u001B[32m";
    private static final String YELLOW = "\u001B[33m";
    private static final String CYAN = "\u001B[36m";

    private static String c(String text, String color) {
        return color + text + RESET;
    }

    private final Scanner scanner;
    private double prob;
    private CoinToss toss;
    private List<String> lastSeries;
    private Stats lastStats;

    public CoinTossApp() {
        scanner = new Scanner(System.in);
        prob = 0.5;
        toss = new CoinToss(prob);
        lastSeries = new ArrayList<>();
        lastStats = new Stats();
    }

    private String ask(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private int askInt(String prompt, int def) {
        while (true) {
            String ans = ask(prompt);
            if (ans.isEmpty()) return def;
            try { return Integer.parseInt(ans); }
            catch (NumberFormatException e) { System.out.println(c("Please enter a number.", YELLOW)); }
        }
    }

    private double askDouble(String prompt, double def) {
        while (true) {
            String ans = ask(prompt);
            if (ans.isEmpty()) return def;
            try { return Double.parseDouble(ans); }
            catch (NumberFormatException e) { System.out.println(c("Please enter a number.", YELLOW)); }
        }
    }

    private void showMenu() {
        System.out.println("\n" + c("═".repeat(50), CYAN));
        System.out.println(c("🪙 COIN TOSS ANALYZER", BRIGHT + CYAN));
        System.out.println(c("═".repeat(50), CYAN));
        System.out.printf("  Probability of Heads: %.2f\n", prob);
        System.out.println(c("═".repeat(50), CYAN));
        System.out.println("  1. 🪙 Single Flip");
        System.out.println("  2. 📊 Series of Flips");
        System.out.println("  3. 📈 Show Statistics");
        System.out.println("  4. 🎯 Set Probability");
        System.out.println("  5. 💾 Export Results");
        System.out.println("  6. 🔁 Continuous Flip Mode");
        System.out.println("  0. 🚪 Exit");
        System.out.println(c("═".repeat(50), CYAN));
    }

    private void singleFlip() {
        String res = toss.flip();
        String color = res.equals("H") ? GREEN : RED;
        System.out.printf("\n🪙 Result: %s (Heads=%.2f)\n", c(res, color), prob);
        lastSeries.add(res);
        lastStats = Analyzer.analyze(lastSeries);
    }

    private void seriesFlips() {
        int n = askInt("Number of flips (default 10): ", 10);
        if (n <= 0) { System.out.println(c("Number must be positive.", RED)); return; }
        System.out.println(c("Flipping " + n + " times...", DIM));
        lastSeries = toss.flipSeries(n);
        lastStats = Analyzer.analyze(lastSeries);
        showStats();
    }

    private void showStats() {
        if (lastStats.total == 0) {
            System.out.println(c("No data yet. Do some flips!", YELLOW));
            return;
        }
        System.out.println("\n📊 STATISTICS");
        System.out.println(c("─".repeat(30), DIM));
        System.out.printf("  Total Flips: %d\n", lastStats.total);
        System.out.printf("  Heads: %d (%.1f%%)\n", lastStats.heads, lastStats.headsPercent);
        System.out.printf("  Tails: %d (%.1f%%)\n", lastStats.tails, lastStats.tailsPercent);
        System.out.printf("  Longest Run (H): %d\n", lastStats.longestRunHeads);
        System.out.printf("  Longest Run (T): %d\n", lastStats.longestRunTails);
        System.out.printf("  Number of Runs: %d\n", lastStats.runs);
        System.out.println("\n📊 Distribution:");
        System.out.println(Analyzer.drawAsciiChart(lastStats, 40));
    }

    private void setProbability() {
        double p = askDouble("Probability of Heads (0.0–1.0): ", 0.5);
        if (p >= 0 && p <= 1) {
            prob = p;
            toss = new CoinToss(p);
            System.out.println(c(String.format("✅ Probability set to %.2f", p), GREEN));
        } else {
            System.out.println(c("Invalid probability.", RED));
        }
    }

    private void exportResults() {
        if (lastStats.total == 0) {
            System.out.println(c("No data to export.", YELLOW));
            return;
        }
        String fname = ask("Filename (leave empty for auto): ");
        if (fname.isEmpty()) fname = null;
        try {
            Exporter.exportResults(lastSeries, lastStats, fname);
        } catch (IOException e) {
            System.out.println(c("❌ Export failed: " + e.getMessage(), RED));
        }
    }

    private void continuousFlip() {
        System.out.println(c("\n🔁 Continuous Flip Mode (press Ctrl+C to stop)", CYAN));
        int count = 0, heads = 0, tails = 0;
        List<String> series = new ArrayList<>();
        while (true) {
            String res = toss.flip();
            series.add(res);
            if (res.equals("H")) heads++; else tails++;
            count++;
            System.out.printf("\r %d flips: H=%d (%.1f%%) T=%d (%.1f%%)", count, heads, (double)heads/count*100, tails, (double)tails/count*100);
        }
    }

    public void run() {
        System.out.print("\033[H\033[2J");
        System.out.flush();
        System.out.println(c("\n🪙 Coin Toss – Advanced Flip Analyzer", BRIGHT + CYAN));
        System.out.println(c("Flip, analyze, and master probability!", DIM));

        while (true) {
            showMenu();
            String choice = ask("Your choice: ");
            switch (choice) {
                case "1": singleFlip(); break;
                case "2": seriesFlips(); break;
                case "3": showStats(); break;
                case "4": setProbability(); break;
                case "5": exportResults(); break;
                case "6": continuousFlip(); break;
                case "0":
                    System.out.println(c("👋 Goodbye!", CYAN));
                    return;
                default:
                    System.out.println(c("❌ Invalid choice.", RED));
            }
            if (!choice.equals("0") && !choice.equals("6")) {
                System.out.print("\nPress Enter to continue...");
                scanner.nextLine();
            }
        }
    }

    public static void main(String[] args) {
        try {
            new CoinTossApp().run();
        } catch (Exception e) {
            System.err.println(c("❌ Unexpected error: " + e.getMessage(), RED));
            e.printStackTrace();
            System.exit(1);
        }
    }
}
